### PR TITLE
Fix regression in 0.6.4 where token store roles could not properly wo…

### PR DESCRIPTION
**Pull request by jefferai**
…rk as a blacklist

### CODE REVIEW

File: vault/token_store.go [hunk 1]
```diff
@@ -1472,7 +1472,11 @@ func (ts *TokenStore) handleCreateCommon(
 			// this is a role, add default unless explicitly disabled.
 			if len(finalPolicies) == 0 {
 				finalPolicies = policyutil.SanitizePolicies(parent.Policies, localAddDefault)
-			} else {
+			} else if len(role.DisallowedPolicies) == 0 {
+				// We only do this if role.DisallowedPolicies is empty because
```
Comment (briankassouf) [inline vault/token_store.go]: Is it even possible to get into this branch? Above the case is:
```go
case role != nil && (len(role.AllowedPolicies) > 0 || len(role.DisallowedPolicies) > 0):
```
So I don't think both `role.AllowedPolicies` and `role.DisallowedPolicies` can be 0.

Comment (jefferai) [inline vault/token_store.go]: That's a good point. I think this is part of the holdover from the major logic change in 0.6.4 that made having both allowed/disallowed not possible.

File: vault/token_store.go [hunk 2]
```diff
@@ -1468,18 +1468,10 @@ func (ts *TokenStore) handleCreateCommon(
 				}
 			}
 		} else {
-			// Check against parent policies, or assign parent policies. As
-			// this is a role, add default unless explicitly disabled.
+			// Assign parent policies if none have been requested. As this is a
+			// role, add default unless explicitly disabled.
 			if len(finalPolicies) == 0 {
 				finalPolicies = policyutil.SanitizePolicies(parent.Policies, localAddDefault)
-			} else {
```
Comment (vishalnayak) [inline vault/token_store.go]: I don't understand why this block is removed. When `allowed_roles` are not specified, not having this block will let a request for any policy to go through. No?

Comment (jefferai) [inline vault/token_store.go]: Yep, that's the documeted design, not allowing this was the regression!

**Comment by dadgar** [comment]
@jefferai Still doesn't work and odd behavior even in released version of Vault if the root policies is or isn't included in the disallowed_policies list.

https://gist.github.com/dadgar/37808dcfdeeffa7f98cbc7b00d44859b

**Pull request merged**